### ISSUE TITLE
Crash in ParserResultsComposed.resolveType(typeName:containingType:)

### ISSUE THREAD

**Issue by fabianmuecke** [krzysztofzablocki/Sourcery#1195]
Hey!

We're seeing a random, but regularly occurring crash, when running Sourcery. We have a large codebase and a very large config, so unfortunately I cannot share the details of this. The crash roughly occurs in 1/4 of runs for us. I do have a [symbolicated crash report](https://github.com/krzysztofzablocki/Sourcery/files/12428631/Sourcery-Crash.txt), which I will attach (and can get your more, if that helps).

Turning on `--serialParse` doesn't help with this, as the `Array.parallelPerform` function is called without regard to that parameter. I have so far never been able to reproduce this with a non-optimized build (it does reproduce with a DEBUG build with `swiftSettings: [SwiftSetting.unsafeFlags(["-cross-module-optimization", "-O"])]` set for all targets, though). I don't know, if this means there might be a compiler optimization issue or if non-optimized builds are simply too slow to reproduce the race condition leading to the crash.

I noticed the problem with 2.0.1, but both 2.0.2 and current master branch show the same behavior.

The crash always seems to happen in either objc_release or objc_retain, called from within `ParserResultsComposed.resolveType(typeName:containingType:)`.

I tried to debug it as far as I could and noticed, it is always trying to resolve the same typealias, when it crashes. We use that typealias in a lot of places and it seems random at which usage it crashes, but it's always been this type, as far as I could see. I cannot share our whole code with you, but I think I can share a renamed version of the type(s) in question.

So here we go:
```swift
enum SomeEnum<T> {
    case one(T)
    case two(T? = nil)
    case error(Error, flag: Bool)

    var value: T? {
        switch self {
        case let .one(value),
             let .two(.some(value)):
            return value
        case .two(.none),
             .error:
            return nil
        }
    }

    var isTwo: Bool {
        switch self {
        case .one, .error:
            return false
        case .two:
            return true
        }
    }
}

typealias SomeTypealias<T> = SomeEnum<OtherEnum<T>>

enum OtherEnum<T> {
    case one(T)
    case two(T)
}

extension OtherEnum: HasValue {
    var value: T {
        switch self {
        case let .one(value):
            return resource
        case let .two(value):
            return value
        }
    }
}

protocol HasValue {
    associatedtype T
    var value: T { get }
}

extension SomeEnum where T: HasValue {
    var content: T.T? {
        value?.value
    }
}
```

None of these types need to perform any Sourcery generation on themselves, but they might be used by other types, for which code is generated, so I cannot simply remove the source file from parsing.

Any help would be appreciated. This bug has increasingly become an issue for us, as we run source code generation as part of our CI pipeline and thus our CI keeps failing randomly.

**Comment by art-divin** [comment]
👋🏻  Hello @fabianmuecke , 

thanks for reporting! I'll have a look as soon as I have a minute ⏲️ Will keep you posted.

In the meantime, you are welcome to investigate & fix this issue yourself, I'll provide as much help as I can 👍🏻

**Comment by fabianmuecke** [comment]
> 👋🏻 Hello @fabianmuecke ,
> 
> thanks for reporting! I'll have a look as soon as I have a minute ⏲️ Will keep you posted.
> 
> In the meantime, you are welcome to investigate & fix this issue yourself, I'll provide as much help as I can 👍🏻

Thank you, @art-divin!

Just wanted to share one more quick detail. I tried to remove the typealias and replace all usages with direct type usage, then had our source code generation run in a loop (which is how I debugged this to begin with) for over 30 minutes now, without failing (each pass takes roughly 7-8 seconds). So I guess it's safe to say: removing the typealias works around the issue. The issue must be related to the typealias somehow.
